src/import/sst_service.rs

```diff
@@ -125,46 +131,46 @@ impl<Router: RaftStoreRouter<RocksEngine>> ImportSst for ImportSSTService<Router
         }
         let label = "upload";
         let timer = Instant::now_coarse();
-        let import = Arc::clone(&self.importer);
-        let bounded_stream = mpsc::spawn(stream, &self.threads, self.cfg.stream_channel_window);
-
-        ctx.spawn(
-            self.threads.spawn(
-                bounded_stream
-                    .into_future()
-                    .map_err(|(e, _)| Error::from(e))
-                    .and_then(move |(chunk, stream)| {
-                        // The first message of the stream contains metadata
-                        // of the file.
-                        let meta = match chunk {
-                            Some(ref chunk) if chunk.has_meta() => chunk.get_meta(),
-                            _ => return Err(Error::InvalidChunk),
-                        };
-                        let file = import.create(meta)?;
-                        Ok((file, stream))
-                    })
-                    .and_then(move |(file, stream)| {
-                        stream
-                            .map_err(Error::from)
-                            .fold(file, |mut file, chunk| {
-                                let start = Instant::now_coarse();
-                                let data = chunk.get_data();
-                                if data.is_empty() {
-                                    return future::err(Error::InvalidChunk);
-                                }
-                                if let Err(e) = file.append(data) {
-                                    return future::err(e);
-                                }
-                                IMPORT_UPLOAD_CHUNK_BYTES.observe(data.len() as f64);
-                                IMPORT_UPLOAD_CHUNK_DURATION.observe(start.elapsed_secs());
-                                future::ok(file)
-                            })
-                            .and_then(|mut file| file.finish())
+        let import = self.importer.clone();
+        let (rx, buf_driver) =
+            create_stream_with_buffer(stream.compat(), self.cfg.stream_channel_window);
+        let mut rx = rx.map_err(Error::from);
+
+        let handle_task = async move {
+            let res = async move {
```

Comment:
Why do we need the inner async block?

Reply:
In this way, we can conveniently use `?` and then process the result uniformly.